### PR TITLE
Add sharding support and fix the world axis

### PR DESCRIPTION
**Pull request by amacati**
## Rationale
We want to support sharding of the simulation so that we can run on multiple devices. Multi-device support unlocks performance improvements on the CPU and might be interesting for future parallel GPU simulations. This PR adds the tools for sharding.

### What was previously broken
While working on this, I noticed that some of our functions are fundamentally broken. This mostly affects resets and sharding. When we reset our data, we previously made a shape check if the array matches the number of worlds in its leading dimension. This leads to wrong positives, i.e. the `gravity_vector` with `shape=(3,)` would get picked up for `n_worlds=3`. This is obviously wrong. There is no way to tell which array just happens to coincide with the simulation n_worlds, and which ones are truly parallel across worlds.

### The world axis
We have two realistic options: either we keep an explicit list of all arrays that are batched over worlds somewhere, or we somehow mark each array. 

The first option feels clunky, since we would store information about the data in two places, and users would have to use a separate registration mechanism for plugins.

The second option needs some way to attach metadata to arrays. Luckily, `flax.struct.field` already has a `metadata` field. We now use this for every array that has a world axis. Fields are marked using `field(metadata={WORLD_INDEXED_KEY: True})`.

While adding this, I noticed that many arrays document false shape information. This is now also fixed.

### Checking which arrays are world-indexed
We have a new utils function, `world_mask`, that maps out (possibly nested) structs. It replicates their structure exactly, but replaces leaves with a bool flag that indicates if they have a per-world axis. For each array with the `WORLD_INDEXED_KEY` flag set, it inserts a True.

### Proper reset policies
The world axis allows us to formulate proper policies around resets. Obviously, masked resets can only reset arrays that are replicated per-world. Shared arrays cannot be decided. The new `reset` function uses a revised version of `pytree_replace()` that takes the `world_mask` PyTree and only resets the proper fields. Importantly, this also applies for all-True masks.

If `None` is passed, we instead reset everything, including the shared data, because the case is trivially decidable, and the behavior is what users would expect.

## Support for sharding across devices
Having introduced all this machinery, sharding, luckily, becomes largely trivial. We shard all world-axis arrays evenly across devices and replicate shared arrays. Selecting world-batched arrays is trivial with the new `world_mask` function.

One caveat is that we currently cannot support sharing in explicit mode. SciPy introduces some scatter operations that cannot be resolved in explicit mode, so we pre-select `jax.sharding.AxisType.Auto` (see https://docs.jax.dev/en/latest/parallel.html#auto-sharding-mode-decides-shardings-automatically-during-compilation). Fixing this will require work upstream in SciPy and is out of scope for now.

Related to efficient sharding: #98

### CODE REVIEW

File: docs/user-guide/sharding.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    Workloads that do not make use of sharding are unaffected by replication etc.
+
+```python
+import jax
+
+from crazyflow.sim import Sim
+from crazyflow.sim.sharding import world_mesh
+
+devices = jax.devices()
+sim = Sim(n_worlds=4 * len(devices))
```
Comment (ratheron) [inline docs/user-guide/sharding.md:15]: Why is there a 4? Probably arbitrary number. Below you say that n_worlds needs to be a multiple of n_devices. Why not just put len(devices) there or add that 4 is just an arbitrary scalar?

File: docs/user-guide/sharding.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+assert sim.data.states.pos.sharding.spec == jax.sharding.PartitionSpec("worlds")
+assert sim.data.params.gravity_vec.sharding.spec == jax.sharding.PartitionSpec()
+```
+
+The number of worlds has to be divisible by the number of devices. Arrays declare that they are batched over worlds on their fields, see [The world axis](world-axis.md). Data shared across worlds, such as the gravity vector above, is replicated on every device.
+
+!!! warning
+    Sharding relies on JAX's automatic sharding mode, which `world_mesh` requests for you. `jax.make_mesh` defaults to explicit sharding, under which the Mellinger controller fails inside SciPy's rotation backend.
+
+## Functional API
```
Comment (ratheron) [inline docs/user-guide/sharding.md:28]: Can we add a sentence to this part what this would be needed for?

File: crazyflow/dynamics/so_rpy_rotor_drag/dynamics.py
Comment (ratheron) [inline crazyflow/dynamics/so_rpy_rotor_drag/dynamics.py:1]: As far as I understand this part, we limit ourselves to only randomize J, J_inv, and m across worlds? This seems limiting for proper DR, no?

File: docs/user-guide/sharding.md (hunk 1 above)
Comment (amacati) [reply docs/user-guide/sharding.md:15]: Arbitrary, yeah. The point was "more than one world per device"

File: docs/user-guide/sharding.md (hunk 2 above)
Comment (amacati) [reply docs/user-guide/sharding.md:28]: You mean why we would want explicit sharding? Yes, we could add that. It adds some restrictions, but makes sharding queriable during trace time: https://docs.jax.dev/en/latest/parallel.html

File: crazyflow/dynamics/so_rpy_rotor_drag/dynamics.py
Comment (amacati) [reply crazyflow/dynamics/so_rpy_rotor_drag/dynamics.py:1]: Yes. I'm not sure how we would turn them into per-world fields. Maybe one can add the metadata when tiling them to `n_worlds`

File: docs/user-guide/sharding.md (hunk 2 above)
Comment (ratheron) [reply docs/user-guide/sharding.md:28]: I meant the functional API part

Comment (amacati) [reply docs/user-guide/sharding.md:28]: Ah, it's just there so that we can work on everything without the `Sim` object. Everything else is the same. As always, the OOP version is implemented using this functional API, so there is no duplication

File: crazyflow/dynamics/first_principles/dynamics.py [hunk 3]
```diff
@@ -298,27 +298,27 @@ def symbolic_dynamics(
 
 @dataclass
 class Params:
-    mass: Array  # (N, M, 1)
+    mass: Array = field(metadata={CORE_NDIM_KEY: 1})  # (N, M, 1)
     """Mass of the drone."""
-    L: Array  # (N, M, 1)
+    L: Array = field(metadata={CORE_NDIM_KEY: 0})  # ()
```
Comment (ratheron) [inline crazyflow/dynamics/first_principles/dynamics.py:303]: In this form it's not possible to simulate uneven propeller effects, since L, prop_inertia, and rotor_dyn are the same for all propellers. In real, degrading motors and props will have affect the dynamics. Fixing this should not be part of this PR, but we should probably open an issue on this to keep it in mind. However, I think allowing randomizing per world and drone should already work by changing the params to one dimensional arrays and setting core ndim to 1.

File: crazyflow/dynamics/so_rpy/dynamics.py [hunk 4]
```diff
@@ -316,31 +316,31 @@ def symbolic_dynamics_euler(
...
-    J: Array  # (N, M, 3, 3)
+    J: Array = field(metadata={CORE_NDIM_KEY: 2})  # (N, M, 3, 3)
     """Inertia matrix of the drone."""
 
-    J_inv: Array  # (N, M, 3, 3)
+    J_inv: Array = field(metadata={CORE_NDIM_KEY: 2})  # (N, M, 3, 3)
     """Inverse of the inertia matrix of the drone."""
 
-    acc_coef: Array  # (N, M, 1)
+    acc_coef: Array = field(metadata={CORE_NDIM_KEY: 0})  # ()
```
Comment (ratheron) [inline crazyflow/dynamics/so_rpy/dynamics.py:331]: Same here as in the first principles model.

File: crazyflow/dynamics/so_rpy_rotor/dynamics.py [hunk 5]
```diff
@@ -375,25 +375,25 @@ def symbolic_dynamics_euler(
...
+    gravity_vec: Array = field(metadata={CORE_NDIM_KEY: 1})  # (3,)
     """Gravity vector of the drone."""
-    J: Array  # (N, M, 3, 3)
+    J: Array = field(metadata={CORE_NDIM_KEY: 2})  # (N, M, 3, 3)
     """Inertia matrix of the drone."""
-    J_inv: Array  # (N, M, 3, 3)
+    J_inv: Array = field(metadata={CORE_NDIM_KEY: 2})  # (N, M, 3, 3)
     """Inverse of the inertia matrix of the drone."""
-    thrust_time_coef: Array  # (N, M, 1)
+    thrust_time_coef: Array = field(metadata={CORE_NDIM_KEY: 0})  # ()
```
Comment (ratheron) [inline crazyflow/dynamics/so_rpy_rotor/dynamics.py:386]: Same as above

File: crazyflow/dynamics/so_rpy_rotor_drag/dynamics.py [hunk 6]
```diff
@@ -409,27 +409,27 @@ def symbolic_dynamics_euler(
...
+    gravity_vec: Array = field(metadata={CORE_NDIM_KEY: 1})  # (3,)
     """Gravity vector of the drone."""
-    J: Array  # (N, M, 3, 3)
+    J: Array = field(metadata={CORE_NDIM_KEY: 2})  # (N, M, 3, 3)
     """Inertia matrix of the drone."""
-    J_inv: Array  # (N, M, 3, 3)
+    J_inv: Array = field(metadata={CORE_NDIM_KEY: 2})  # (N, M, 3, 3)
     """Inverse of the inertia matrix of the drone."""
-    thrust_time_coef: Array  # (N, M, 1)
+    thrust_time_coef: Array = field(metadata={CORE_NDIM_KEY: 0})  # ()
```
Comment (ratheron) [inline crazyflow/dynamics/so_rpy_rotor_drag/dynamics.py:420]: Same as above

File: crazyflow/sim/data.py [hunk 7]
```diff
@@ -205,19 +207,17 @@ def create(
 class SimCore:
     freq: int = field(pytree_node=False)
     """Frequency of the simulation."""
-    device: Device = field(pytree_node=False)
```
Comment (ratheron) [inline crazyflow/sim/data.py:208]: The device is gone because of sharding?

File: docs/user-guide/world-axis.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+assert not mask.params.drag_matrix  # (3, 3), shared by all worlds
+```
+
+`world_mask` returns a pytree of booleans matching the data, with one flag per array. Arrays whose field declares no core ndim are shared by all worlds.
+
+## Resets
+
+`sim.reset()` restores the whole simulation, both the world-indexed and the shared arrays, with the rng key as the only exception.
+
+`sim.reset(mask)` is the constrained form. It selects worlds along the world axis, so it can only restore arrays that have one. A shared array cannot be restored for some worlds and not others, so a masked reset leaves it untouched:
```
Comment (ratheron) [inline docs/user-guide/world-axis.md:30]: Just for my understanding. Normally, we would just do a normal reset which will reset everything including shared variables such as gravity. If we change the gravity in one world, it's changed in all worlds since it's not batched. A reset in one world would then not trigger a reset of non shared variables, right? I can't think of an example where the second case realistically occurs

File: crazyflow/dynamics/first_principles/dynamics.py (hunk 3 above)
Comment (amacati) [reply crazyflow/dynamics/first_principles/dynamics.py:303]: Yes, but I would change the dimension when it arises, not now. Everything else would be confusing

File: crazyflow/dynamics/so_rpy/dynamics.py (hunk 4 above)
Comment (amacati) [reply crazyflow/dynamics/so_rpy/dynamics.py:331]: I would address this in https://github.com/learnsyslab/crazyflow/issues/100, not here. It would be confusing

File: crazyflow/dynamics/so_rpy_rotor/dynamics.py (hunk 5 above)
Comment (amacati) [reply crazyflow/dynamics/so_rpy_rotor/dynamics.py:386]: Deferred to #100

File: crazyflow/dynamics/so_rpy_rotor_drag/dynamics.py (hunk 6 above)
Comment (amacati) [reply crazyflow/dynamics/so_rpy_rotor_drag/dynamics.py:420]: Deferred to #100

File: crazyflow/sim/data.py (hunk 7 above)
Comment (amacati) [reply crazyflow/sim/data.py:208]: Yes and no. With sharding, we have no unique definition of device. Some arrays are copied over all devices, some are sharded. The device was also never necessary in the first place. It was only used in one place where we wanted to cast input explicitly into jax arrays. This is not compatible with sharding, and there are much cleaner solutions for it.

File: docs/user-guide/world-axis.md (hunk 8 above)
Comment (amacati) [reply docs/user-guide/world-axis.md:30]: Correct. And yes, it is unlikely to happen, but we spell it out to make it absolutely clear

File: crazyflow/dynamics/so_rpy_rotor_drag/dynamics.py
Comment (amacati) [reply crazyflow/dynamics/so_rpy_rotor_drag/dynamics.py:1]: Resolved with the `core_ndim` approach

**Review by ratheron — APPROVED** [review]
